packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,240 @@
+
+import type { ConfigRoute, RouteManifest } from "./routes";
+import {
+  escapeEnd,
+  escapeStart,
+  optionalEnd,
+  optionalStart,
+  paramPrefixChar,
+  routeModuleExts,
+} from "./routesConvention";
+
+export function flatRoutes(
+  appDirectory: string,
+  ignoredFilePatterns?: string[]
+): RouteManifest {
+  let extensions = Array.from(routeModuleExts).join(",");
+  let routePaths = fg.sync(`*.${extensions}`, {
+    absolute: true,
+    cwd: path.join(appDirectory, "routes"),
+    ignore: ignoredFilePatterns,
+  });
+
+  return flatRoutesUniversal(appDirectory, routePaths);
+}
+
+/**
+ * Create route configs from a list of routes using the flat routes conventions.
+ * @param {string} appDirectory The absolute root directory the routes were looked up from.
+ * @param {string[]} routePaths The absolute route paths.
+ * @param {string} [prefix = "routes"] The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+) {
+  let routes: Record<string, ConfigRoute> = {};
+  let sortedRoutes = routePaths
+    .sort((a, b) => (a.length - b.length > 0 ? 1 : -1))
+    .map((routePath) =>
+      routePath.slice(appDirectory.length + 1).replace(/\\/g, "/")
+    );
+  let processedIds: string[] = [];
+  for (let routePath of sortedRoutes) {
+    let routeId = routeIdFromPath(routePath);
+    if (routes[routeId]) {
+      throw new Error(`Duplicate route: ${routeId}`);
+    }
+
+    let parentId = "";
+    for (let processedId of processedIds) {
+      if (routeId.startsWith(processedId.replace(/\$$/, "*"))) {
+        parentId = processedId;
+        break;
+      }
+    }
+
+    routes[routeId] = {
+      id: routeId,
+      path: pathFromRouteId(routeId, parentId || prefix),
+      parentId: parentId || "root",
+      file: routePath,
+    };
+    if (isIndexRoute(routeId)) {
+      routes[routeId].index = true;
+    }
+    processedIds.unshift(routeId);
+  }
+
+  return routes;
+}
+
+function routeIdFromPath(relativePath: string) {
+  return (
+    relativePath
+      .split(".")
+      // remove file extension
+      .slice(0, -1)
+      .join(".")
+  );
+}
+
+export function pathFromRouteId(routeId: string, parentId: string) {
+  let parentPath = "";
+  if (parentId) {
+    parentPath = getRouteSegments(parentId, true)[0].join("/");
+  }
+  if (parentPath.startsWith("/")) {
+    parentPath = parentPath.substring(1);
+  }
+  let routePath = getRouteSegments(routeId, true)[0].join("/");
+  if (routePath.startsWith("/")) {
+    routePath = routePath.substring(1);
+  }
+  let pathname = parentPath
+    ? routePath.slice(parentPath.length + 1)
+    : routePath;
+  if (pathname.endsWith("/_index")) {
+    pathname = pathname.replace(/_index$/, "");
+  }
+  if (pathname.startsWith("/")) {
+    pathname = pathname.substring(1);
+  }
+  return pathname || undefined;
+}
+
+function isIndexRoute(routeId: string) {
+  return routeId.endsWith("_index");
```

Comment:
We had [discussed using just `_` as the index route convention](https://github.com/remix-run/remix/discussions/4482#discussioncomment-4052865) and last time I spoke with @ryanflorence we agreed on it.

Could we change this to use `_` for index instead?
Note: this might mean we need more robust checks for pathless/layout routes since `endsWith('_')` would then incorrectly capture index routes.

Reply:
last i spoke with @ryanflorence (last wednesday) we decided `_index`, but happy to adjust if that's since changed

Reply:
Spoke to @ryanflorence today. Going with `_index`.